# Titanic Survival

#### Grading:


- Code: 90 pts
- Markdown Documentation (Documentation within IPython using Markdown): 10 pts

## Due: 04/03/2023


We are going to study the survival rate of passengers on titanic and what variables affected survival.

Load the dataset in `titanic.xls`. It contains data on all the passengers that travelled on the Titanic.

In [168]:
from IPython.core.display import HTML
import pandas as pd

HTML(filename='../data/titanic.html')

In [169]:
# you would need xlrd - pip install xlrd
t_file = pd.ExcelFile('../data/titanic.xls')
t_df = t_file.parse("titanic", header=None)
t_df.head()

0         1                                     2       3       4   \
0  pclass  survived                                  name     sex     age   
1       1         1         Allen, Miss. Elisabeth Walton  female      29   
2       1         1        Allison, Master. Hudson Trevor    male  0.9167   
3       1         0          Allison, Miss. Helen Loraine  female       2   
4       1         0  Allison, Mr. Hudson Joshua Creighton    male      30   

      5      6       7         8        9         10    11    12  \
0  sibsp  parch  ticket      fare    cabin  embarked  boat  body   
1      0      0   24160  211.3375       B5         S     2   NaN   
2      1      2  113781    151.55  C22 C26         S    11   NaN   
3      1      2  113781    151.55  C22 C26         S   NaN   NaN   
4      1      2  113781    151.55  C22 C26         S   NaN   135   

                                13  
0                        home.dest  
1                     St Louis, MO  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

### Women and children first?

*** 1. Use the `groupby` method to calculate the proportion of passengers that survived by sex. (25 pts)*** 

In [170]:
t_df.columns = t_df.loc[0]
t_df = t_df.drop(t_df.index[0]).reset_index().drop('index', axis=1)
t_df['survived'] = t_df.survived.astype('float')
survived_count = t_df[t_df['survived'] == 1].shape[0]
t_df.head(3)

0 pclass  survived                            name     sex     age sibsp  \
0      1       1.0   Allen, Miss. Elisabeth Walton  female      29     0   
1      1       1.0  Allison, Master. Hudson Trevor    male  0.9167     1   
2      1       0.0    Allison, Miss. Helen Loraine  female       2     1   

0 parch  ticket      fare    cabin embarked boat body  \
0     0   24160  211.3375       B5        S    2  NaN   
1     2  113781    151.55  C22 C26        S   11  NaN   
2     2  113781    151.55  C22 C26        S  NaN  NaN   

0                        home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON

* First things first, I have set the columns to be the first row, as it has been ambigous with a numbered columns and a numbered index.
After that, index reset, to restore the records indices to start from 0.
* Converting _survived_ column to 'float' so arithmatic operations could make sense on it
* Calculated survived count to use in ratios calculation

In [171]:
t_df.groupby('sex').sum(numeric_only=True)['survived']/survived_count

sex
female    0.678
male      0.322
Name: survived, dtype: float64

* Group by is used to group records by _sex_, using summation, to get the total survived number, and then dividing by the pre-calculated _survived_count_ to get the ratio of each sex survival rate

### Findings
* 68% of the survivors were Females
* 32% of the survivors were Males

*** 2. Calculate the same proportion, but by class and sex. (25 pts)*** 

In [172]:
t_df.groupby(['sex','pclass']).sum(numeric_only=True)['survived']/survived_count

sex     pclass
female  1         0.278
        2         0.188
        3         0.212
male    1         0.122
        2         0.050
        3         0.150
Name: survived, dtype: float64

* Group by is used to group records by _sex_ and _pclass_, using summation, to get the total survived number, and then dividing by the pre-calculated _survived_count_ to get the ratio of each sex survival rate

### Findings
* Females of first class are the highest in survival rate, with a percentage of 27.8% of the surviving number.
* Comes in second place the females of the third classs with a percentage of 21.2%.
* Highest Male surviving class is still below the lowest female surviving class, with a percentage of 15% and 18.7% respectively.


*** 3. Create age categories: children (under 14 years), adolescents (14-20), adult (21-64), and senior(65+), and calculate survival proportions by age category, class and sex. (40 pts)***

In [173]:
t_df['age_group'] = pd.cut(x=t_df['age'], bins=[0, 14, 20, 64, 200],
                     labels=['children', 'adolescents', 'adult',
                             'senior'])

* Using Pandas _cut_ to generate categorical values for the age groups in a new column called _age_group_

In [174]:
t_cat = t_df.groupby(['age_group', 'pclass', 'sex']).sum(numeric_only=True)['survived']/survived_count
t_cat

age_group    pclass  sex   
children     1       female    0.028
                     male      0.022
             2       female    0.036
                     male      0.024
             3       female    0.042
                     male      0.044
adolescents  1       female    0.040
                     male      0.006
             2       female    0.024
                     male      0.004
             3       female    0.040
                     male      0.018
adult        1       female    0.186
                     male      0.076
             2       female    0.124
                     male      0.018
             3       female    0.062
                     male      0.056
senior       1       female    0.002
                     male      0.002
             2       female    0.000
                     male      0.000
             3       female    0.000
                     male      0.000
Name: survived, dtype: float64

* Group by is used to group records by _sex_ , _pclass_, and _age_group_, using summation, to get the total survived number, and then dividing by the pre-calculated _survived_count_ to get the ratio of each sex survival rate

### Findings

* Nearly 0% of the surviving were seniors, however, the surviving <0% are all from the first class.
* Adults are the highest in surviving rates, with the higher surviving sex being females.
* Comes in second place children with a percentage of 19.6%.